lib/bech32/test/Codec/Binary/Bech32Spec.hs

```diff
@@ -124,104 +126,132 @@ spec = do
 
         it "Decoding fails when an adjacent pair of characters is swapped." $
             property $ \s -> do
-                let validString = getValidBech32String s
-                index <- choose (0, T.length validString - 2)
-                let prefix = T.take index validString
-                let suffix = T.drop (index + 2) validString
-                let char0 = T.singleton (T.index validString index)
-                let char1 = T.singleton (T.index validString $ index + 1)
-                let recombinedString = prefix <> char1 <> char0 <> suffix
-                return $
-                    (T.length recombinedString === T.length validString)
-                    .&&.
-                    (Bech32.decode recombinedString `shouldSatisfy`
-                        (if char0 == char1 then isRight else isLeft))
+                let originalString = getValidBech32String s
+                index <- choose (0, T.length originalString - 2)
+                let prefix = T.take index originalString
+                let suffix = T.drop (index + 2) originalString
+                let char1 = T.singleton (T.index originalString index)
+                let char2 = T.singleton (T.index originalString $ index + 1)
+                let corruptedString = prefix <> char2 <> char1 <> suffix
+                let description = intercalate "\n"
+                        [ "index of char #1: " <> show index
+                        , "index of char #2: " <> show (index + 1)
+                        , "         char #1: " <> show char1
+                        , "         char #2: " <> show char2
+                        , " original string: " <> show originalString
+                        , "corrupted string: " <> show corruptedString ]
+                return $ counterexample description $
+                    char1 /= char2 ==>
```

Comment:
Could this precondition be removed? Then change then line below to `shouldSatisfy (if char1 == char2 then isRight else isLeft)`

Reply:
@rvl wrote:

> Could this precondition be removed? Then change then line below to `shouldSatisfy (if char1 == char2 then isRight else isLeft)`

Actually, we do want this precondition, as we're **only** interested in the cases where corruption actually does happen. By making corruption a precondition, we guarantee to produce cases that we're interested in (or else QuickCheck will give up), where decoding is presumed to fail.

As it turns out, only a small proportion of tests (those that fail to produce a corrupted string) are discarded here.

Though perhaps you have another reason in mind for removing the precondition?